tools/isobuild/bundler.js

```diff
@@ -3181,80 +3138,92 @@ function bundle({
-    // another solution eventually (probably some kind of mount
-    // directive that mounts the client bundle in the server at runtime)
-    var getRelativeTargetPath = function (options) {
-      var pathForTarget = function (target) {
-        var name;
-        _.each(targets, function (t, n) {
-          if (t === target) {
-            name = n;
-          }
-        });
-        if (! name) {
-          throw new Error("missing target?");
-        }
-        return files.pathJoin('programs', name);
-      };
-
-      return files.pathRelative(pathForTarget(options.relativeTo),
-                                pathForTarget(options.forTarget));
-    };
+    const targets = Object.create(null);
+    const hasOwn = Object.prototype.hasOwnProperty;
 
     // Write to disk
-    var writeOptions = {
+    const writeOptions = {
       includeNodeModules,
       builtBy,
       releaseName,
-      getRelativeTargetPath,
-      minifyMode: minifyMode
+      minifyMode,
     };
 
+    function writeClientTarget(target) {
+      const { arch } = target;
+      const written = writeTargetToPath(arch, target, outputPath, {
+        buildMode: buildOptions.buildMode,
+        previousBuilder: previousBuilders[arch],
+        ...writeOptions,
+      });
+      clientWatchSet.merge(target.getWatchSet());
+      previousBuilders[arch] = written.builder;
+    }
+
+    // Client
+    webArchs.forEach(arch => {
+      if (allowDelayedClientBuilds &&
+          hasOwn.call(previousBuilders, arch) &&
+          projectContext.platformList.canDelayBuildingArch(arch)) {
+        // If delayed client builds are allowed, and we have a previous
+        // builder for this arch, and it's an arch that we can safely
+        // build later (e.g. web.browser.legacy), then schedule it to be
+        // built after the server has started up.
+        postStartupCallbacks.push(async ({
+          pauseClient,
+          refreshClient,
+          runLog,
+        }) => {
+          const start = +new Date;
+
+          // Build the target first.
+          const target = makeClientTarget(app, arch, { minifiers });
+
+          // Tell the webapp package to pause responding to requests from
+          // clients that use this arch, because we're about to write a
+          // new version of this bundle to disk.
+          await pauseClient(arch);
+
+          // Now write the target to disk. Note that we are rewriting the
+          // bundle in place, so this work is not atomic by any means,
+          // which is why we needed to pause the client.
+          writeClientTarget(target);
+
+          // Refresh and unpause the client, now that writing is finished.
+          await refreshClient(arch);
+
+          // Let the webapp package running in the child process know it
+          // should regenerate the client program for this arch.
+          runLog.log(`Finished delayed build of ${arch} in ${
+            new Date - start
+          }ms`, { arrow: true });
```

Comment:
Need to think about whether this logging should remain in the final version. Alternatively, we could add more information about building other bundles, so that this isn't the only message.